0.8 preview

ssorallen opened the pull request:
0.8-preview-merger is a throw away branch based off master. It's here to create a pull request on GitHub for comments.

This is a port of `core-drawer-panel` from 0.5 to 0.8

Notably missing:
- Touch support is still TBD in 0.8, and so the touch-related code here is the same as it was in 0.5.

Commit:
Covert to 0.8 syntax (mostly)

Commit:
Correct style syntax with parentheses

Commit:
Use proper `attrForSelected` attire for core-selector

Commit:
Enable media query changes with callback

Commit:
Move all defaults into `configure`

Commit:
Listen for 'click' delegation to toggle drawer

Commit:
Target non-shadow selector to parent #main container

Commit:
Add demo dependencies as devDeps

Commit:
Add missing repo config pointing to current repo

Commit:
Point Bower config to main HTML file

Commit:
Add missing authors Bower config

Commit:
Match 'Peek' instead of 'Peak' (ain't no mountains)

`startEdgePeek` uses the proper spelling, make the stop method match.

Addresses #52.

Commit:
Strict; `class-name` and more succinct class calc

* Add `"use strict"` scope around the entire element
* Prefer setting `class-name` prop to the `class$` attr because it's
  faster
* Import basic "classNames" (formerly "tokenList") filter from 0.5 to
  succinctly handle class name computation

TODO: Move `classNames` to its own module?

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+  <template>
+    <core-media-query query="[[mediaQuery]]" on-query-matches-changed="onQueryMatchesChanged"></core-media-query>
+
+    <core-selector class-name="[[coreSelectorClass]]" attrForSelected="id" selected="{{selected}}">
```
Do I spy a two-way binding? :sunglasses:

ssorallen replied:
I'd be happy to ditch it, but this is the type of case where the core team seemed to prefer it.

cdata replied:
No worries, I was just interested to hear about it.

ssorallen replied:
:eyes: https://github.com/Polymer/core-drawer-panel/commit/af511823e0bf573b0c8a8fe4b01cd634328561d1

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
-  <core-selector class="{{ {'narrow-layout' : narrow, transition : transition, dragging : dragging, 'right-drawer': rightDrawer} | tokenList }}" valueattr="id" selected="{{selected}}">
+      <div id="main" style="[[mainStyle]]">
+        <content select="[main]"></content>
+        <div id="scrim" on-tap="[[togglePanel]]"></div>
```
What is a scrim? Just curious..

ssorallen replied:
The grayish background that obscures content and on which modal dialogs stack. Not sure the origin of the term, but that's how I've always heard it used.

frankiefu replied:
Should that be ```<div id="scrim" on-click="togglePanel"></div>```?

ssorallen replied:
Ugh, this is one of the places where `core-selector`'s internal `selected` state hides bugs. Even after removing the scrim entirely, clicking the main area toggled the selected panel without being apparent why it was happening. Even one-way binding `[[selected]]` in core-selector didn't change the behavior because core-selector makes changes regardless of how you bind.

ssorallen replied:
I'd like to add `activateEvent=""` to the `core-selector` here to prevent its behavior entirely and make the `onClick` event on the scrim actually do the work. @frankiefu, what do you think of that?

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
-      <content select="[drawer]"></content>
-    </div>
+    </core-selector>
+  </template>
```
Overall this template is way more readable that the 0.5 incarnation :100:

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
-       * @default '640px'
-       */
-      responsiveWidth: '640px',
+    function classNames(obj) {
```
I made the same comment in core-toolbar: can we put this somewhere centralized / shareable?

ssorallen replied:
Yup, I'll find a spot for it today.

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
-      <div id="edgeSwipeOverlay" hidden?="{{!narrow || disableEdgeSwipe}}"></div>
-    </div>
+      <div id="drawer" style="[[drawerStyle]]">
+        <content select="[drawer]"></content>
```
Hrm. I kind of like selecting attributes. Wait, is this the same as liking attribute-based styling? :confused:

ssorallen replied:
I see this as a good use case even though the `select` attribute uses CSS selector syntax. I would prefer to prefix these with `data-` though since `drawer` and `main` are invalid attributes on native elements.

```html
<core-drawer-panel>
  <div data-main></div>
  <div data-drawer></div>
</core-drawer-panel>
```

cdata replied:
:+1:

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+      },
+
+      listeners: {
+        click: 'onClick'
```
Hrm. At one point, @morethanreal told me that we use `*Action` for event handler names. There seems to be some dissonance on this topic in the style guide. Thoughts?

ssorallen replied:
Several people commented on the `Action` suffix and said a prefix is easier to pick out when debugging. I ran it by @morethanreal, and she was on board. I changed this in the style guide too.

cdata replied:
Gah, now I have to go back and change the things I changed! :(

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+
+      // TODO(ssorallen): Convert to 0.8 touch events
+      eventDelegates: {
+        trackstart: 'trackStart',
```
What does `track` mean?

ssorallen replied:
Yeah these I haven't touched. I'm not sure yet.

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+      },
+
+      // Whether the transition is enabled.
+      transition: false,
```
Any particular reason why these aren't initialized via `configure`?

ssorallen replied:
Good call.

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+      },
+
+      ready: function() {
+        this.hasTransform = 'transform' in this.style;
```
Ditto - why not `configure`?

ssorallen replied:
I didn't think about that, I'll try it.

ssorallen replied:
Is the instance available in `configure`? I'm not sure `this.style` is guaranteed to exist yet.

cdata replied:
The instance is available, and `this.style` _appears_ to be available, although I can't say for sure if there are any caveats or not. Give it a shot?

ssorallen replied:
This seems odd to have in this element; this is more the work of something like Modernizr. I can see many elements doing their own feature tests.

cdata replied:
Good point. Maybe remove?

ssorallen replied:
I'll toss it into `configure` for now and think about a feature test util if these pop up in more elements.

cdata commented on index.html:
```diff
@@ -8,15 +8,15 @@
 
-  <script src="../webcomponentsjs/webcomponents.js"></script>
-  <link rel="import" href="../core-component-page/core-component-page.html">
+    <script src="../webcomponentsjs/webcomponents.js"></script>
```
Not `webcomponents-lite.js`?

ssorallen replied:
:eyes:

cdata commented on demo.html:
```diff
@@ -10,59 +10,57 @@
     <meta name="apple-mobile-web-app-capable" content="yes">
-    
+
     <script src="../webcomponentsjs/webcomponents.js"></script>
```
`webcomponents-lite.js`?

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
-    trackStart : function(e) {
-      if (this.swipeAllowed()) {
-        this.dragging = true;
+        this.setAttribute('touch-action', this.swipeAllowed() ? 'pan-y' : '');
```
Even though this isn't computed, perhaps we can write an accessor for the value?

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
-          this.dragging = this.peeking || this.isEdgeTouch(e);
-        }
+      forceNarrowChanged: function() {
+        this.queryMatchesChanged();
```
This method does not seem to exist. Can we add a test for forcing narrow changed?

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+      startEdgePeek: function() {
+        this.width = this.$.drawer.offsetWidth;
+        this.moveDrawer(this.translateXForDeltaX(this.rightDrawer ?
+            -this.edgeSwipeSensitivity : this.edgeSwipeSensitivity));
```
Recognizing this isn't your code, but..

I tend to prefer breaking out ternaries into conditionals or embedding them in discrete accessors. I find that nesting them inline makes them harder to reason about. What do you think?

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+
+      onClick: function(e) {
+        if (e.target && this.toggleAttribute &&
+            e.target.hasAttribute(this.toggleAttribute)) {
```
Maybe assign the result of this boolean logic to a temp that literally describes its meaning?

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+        return !this.disableEdgeSwipe && this.swipeAllowed() &&
+          (this.rightDrawer ?
+            e.pageX >= this.offsetWidth - this.edgeSwipeSensitivity :
+            e.pageX <= this.edgeSwipeSensitivity);
```
Oh my..

ssorallen replied:
Yeah... I haven't touched the touch code yet. :fire:

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
 
-    trackx : function(e) {
-      if (this.dragging) {
+      isMainSelected: function() {
```
We don't have a set convention for preferring accessors or methods but.. uses of this might read a little more nicely if it was an accessor. What are your thoughts?

cdata commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,425 @@
+        if (this.dragging) {
+          this.dragging = false;
+          this.transition = true;
+          this.moveDrawer(null);
```
I'm not totally sure what to expect from passing `null` here. I had to trace through the call to get it. Not sure what to do about that, but felt it was worth mentioning.

ssorallen replied:
:+1: All of the `track` methods are yet-to-be-used. I will revisit your comments when I look at them.

cdata commented:
Looks like there is some work needed still to get touch interactions working. Also, I think we need to add some tests (even if we are only testing the basics for now).

Commit:
Prefer 'lite' version of webcomponents.js

Commit:
Serialize `narrow` as a Boolean as intended

Commit:
Call proper method when `forceNarrow` changes

The method was renamed, but this call location was missed. Need tests...

Commit:
Move all initial values to API method `configure`

This gives a central location for initial values rather than assigning
at instantiation time and then later at configure time.

frankiefu commented on core-drawer-panel.html:
```diff
@@ -75,344 +75,429 @@
-          this.dragging = this.peeking || this.isEdgeTouch(e);
-        }
+      forceNarrowChanged: function() {
+        this.onQueryMatchesChanged();
```
`onQueryMatchesChanged` expects `event` is in the arguments and calls to `event.detail.value`

ssorallen replied:
Thanks, nice catch. This element needs tests.

Commit:
Remove metadata, which is unused in 0.8

Commit:
Normalize interface to `selected` via drawerOpen/Close

Rather than set `selected` in various places, use the drawer open/close
methods to provide a single interface to `selected` and to give semantic
intention to call locations to `selected`.

Commit:
Handle select modification in host

* `scrim` doesn't receive a click event because it's handled by the
  `core-selector` element; no need for another listener there
* Set `selected` based on activate event from the nested
  `core-selector` so all mutation is visible in this file

Commit:
Handle forced narrow layout

Commit:
Normalize to 'e' for event handler arg name

Commit:
Extract onClick val to variable for clarity

Commit:
Move initial values to 'configure'

ssorallen commented:
Closing this as the initial state. I will create future 0.8 changes on branches off the 0.8-preview branch and send PRs to it. Near future work includes touch events once they land in 0.8.
